### ISSUE TITLE
Feature Request: Get all items and search local storage

### ISSUE THREAD

**Issue by rajd90** [cyrilletuzi/angular-async-local-storage#45]
Can you add a way to get all items, so that we can search local storage?

Get all items could be along the lines of.....
        let indexedDB = window.indexedDB
        let open = indexedDB.open("ngStorage");
        let _self = this;
        open.onsuccess = function() {
            let db = open.result;
            let tx = db.transaction("localStorage", "readwrite");
            let store = tx.objectStore("localStorage");
            let pbjects= [];
            store.openCursor().onsuccess = function(event) {
                var cursor = event.target.result;
                if (cursor) {
                    objects.push(cursor.value);
                    cursor.continue();
                }
                else {
                    console.log(objects);
                }
            };

        }        

The above has been adapted from my code (which works)

Once this is in place, it would be possible for us to then filter the array of objects how we want. Right now the only way to get an item is with the key, which makes a search of the store impossible. If you have a better or more efficient way to retrieve all the items from store and search them then even better.

**Comment by cyrilletuzi** [comment]
Problem is: this kind of `IndexedDb`-specific features would not work in the scenarios where the lib has to fallback to `localStorage`.

Anyway, t's not the goal of this lib to mirror all the features of `IndexedDb`. This lib as the same goal as the `localForage` lib: simple API like `localStorage`, but efficient via `IndexedDb` storage when possible. If you have an advanced need of `IndexedDb`, there are already good libraries which simplify the native API.

However, another solution for your problem would just be to store an array. Then you could filter this array.

I'll leave this issue open for now if someone has an idea on how to implement this feature with the constraints I talked about, but for now I don't see a way.

**Comment by rajd90** [comment]
@cyrilletuzi...
Thanks for the speed response. It took quite a bit of rework but i've done as you suggested and stored the entire array in a single object in the store. Thanks 👌🏽

Given how common the scenario of storing an array of objects is likely to be perhaps there is something we can still look at. If searching through the store would be difficult to implement when falling back to different storage types, perhaps there is a way to implement store, search and update of the array in a single object.

Maybe, when storing/setting the item we can give it a type of array/list. And then when getting the item we can optionally specify and ID or a property and search criteria. This saves everyone implementing the logic from scratch each time. It could also help when updating a single item/object in the array

**Comment by fmavdb** [comment]
> Problem is: this kind of `IndexedDb`-specific features would not work in the scenarios where the lib has to fallback to `localStorage`.
> 
> Anyway, it's not the goal of this lib to mirror all the features of `IndexedDb`. This lib as the same goal as the `localForage` lib: simple API like `localStorage`, but efficient via `IndexedDb` storage when possible. If you have an advanced need of `IndexedDb`, there are already good libraries which simplify the native API.
> 
> However, another solution for your problem would just be to store an array. Then you could filter this array.
> 
> I'll leave this issue open for now if someone has an idea on how to implement this feature with the constraints I talked about, but for now I don't see a way.

You can actually get all the keys/values from localstorage rather simple:
```
var keys = Object.keys(localStorage); 
```

Then your able to fill in a map with key and get value with `localStorage.getItem(key)`.

For Indexeddb you can get all keys/values with the function posted above.

**Comment by cyrilletuzi** [comment]
@fmavdb Thanks for the info. It was spotted in another issue as well.

But I'm still concerned by several things:
- when in scenarios when the `localStorage` fallback is in charge, it could hurt performance a lot,
- is it the role of this lib?
- if so, getting all items stored would not be very accurate, what is really wanted are subsequent operations like searching. And if we go this way, it should be designed carefully. It should be consistent with other requests like #39.

I'm open to spec propositions (inspiration from existing IndexedDb libs would be a good idea).

**Comment by cyrilletuzi** [comment]
@rajd90 @fmavdb This will be tracked in #54. Please go to this issue and comment with any useful information. Thanks.

**Comment by cyrilletuzi** [comment]
Added in 7.1.0-beta.0. See #54 to track the stable release.
